va/va_prot.h

```diff
@@ -0,0 +1,364 @@
+ * update_buff.input.data_size = sizeof(info);
+ * va_status = vaCreateBuffer(
+ *                 va_dpy,
+ *                 protected_session,
+ *                 (VABufferType) VAProtectedSessionHwUpdateBufferType,
+ *                 sizeof(update_buff),
+ *                 1,
+ *                 &update_buff,
+ *                 &buffer);
+ *
+ * va_status = vaProtectedSessionHwUpdate(va_dpy, protected_session, buffer);
+ *
+ * vaDestroyBuffer(va_dpy, buffer);
+ * \endcode
+ *
+ * \section api_prot_exec TEE communication via vaProtectedSessionExecute()
+ *
+ * A function for communicating with TEE directly (Trusted Execution Environment). It needs to
+ * create a VA buffer first with type VAProtectedSessionExecuteBufferType, the structure of
+ * this type is IHV's implementation specific. Usually it will have a function_id to specify
+ * which function to execute. The following pseudo-code demonstrates getting session id via
+ * vaProtectedSessionExecute() as an example.
+ *
+ * \code
+ * uint16_t version = 0xFF;
+ * VABufferID buffer;
+ * VAProtectedSessionExecuteBuffer exec_buff = {0};
+ *
+ * exec_buff.function_id = GET_TEE_VERSION
+ * exec_buff.input.input_data = nullptr;
+ * exec_buff.input.data_size = 0;
+ * exec_buff.output.output_data = &version;
+ * exec_buff.output.max_data_size = (uint16_t);
+ * va_status = vaCreateBuffer(
+ *                 va_dpy,
+ *                 protected_session,
+ *                 (VABufferType) VAProtectedSessionExecuteBufferType,
+ *                 sizeof(exec_buff),
+ *                 1,
+ *                 &exec_buff,
+ *                 &buffer);
+ *
+ * va_status = vaProtectedSessionExecute(va_dpy, protected_session, buffer);
+ *
+ * vaDestroyBuffer(va_dpy, buffer);
+ * \endcode
+ */
+
+/**
+ * ProtectedSessions and Contexts
+ *
+ * Context represents a "virtual" video decode, encode or video processing
+ * pipeline. Surfaces are render targets for a given context. The data in the
+ * surfaces are not accessible to the client except if derived image is supported
+ * and the internal data format of the surface is implementation specific.
+ *
+ * A protected session is a session between application and GPU hardware.
+ * Protected sessions can be attached to display to enable protection in the pipeline.
+ *
+ * Both contexts and protected sessions are identified by unique IDs and its
+ * implementation specific internals are kept opaque to the clients
+ */
+typedef VAContextID VAProtectedSessionID;
+
+/** \brief Input buffer of VAProtectedSessionExecuteBuffer or VAProtectedSessionHwUpdateBuffer */
+typedef struct _VAProtectedSessionBufferIn
+{
+    /** \brief Contains the size of the input data. */
+    uint32_t    data_size;
+    /** \brief The input data. */
+    void        *input_data;
+} VAProtectedSessionBufferIn;
+
+/** \brief Output buffer of VAProtectedSessionExecuteBuffer or VAProtectedSessionHwUpdateBuffer */
+typedef struct _VAProtectedSessionBufferOut
```

Comment:
Same comment as before: `VAProtectedSessionBufferIn` and `VAProtectedSessionBufferOut` are effectively the same and can be combined under single `VAProtectedBuffer`. Your use case where you have both input and output in any case have capability to distinguish them by name:
```
typedef struct _VAProtectedSessionExecuteBuffer {
    /** \brief Specify the function to execute. It is IHV's implementation specific */
    uint32_t                        function_id;
    VAProtectedBuffer       input;
    VAProtectedBuffer       output;
    ...
};
```

Reply:
Thanks for your suggestion. I merged VAProtectedSessionBufferIn & VAProtectedSessionBufferOut  into one buffer type VAProtectedSessionBuffer.